### ISSUE TITLE
[Question] Dungeon Strategies - How would you implement them?

### ISSUE THREAD

**Issue by noisiver** [liyunfan1223/mod-playerbots#172]
I'm looking at implementing some strategies for Wrath of the Lich King dungeons, mainly the heroic versions. I'm a bit stumped though as to where to begin. I don't know if it should be done like the strategies for Naxxramas are implemented or not. It seems a bit too complex, I would assume something simpler would be the way to go for dungeons.

I'm curious about the table named `playerbots_dungeon_suggestion_strategy` too - what is it and how does it work? Could it be used to do this?

My C++ isn't fantastic but I'm willing to learn, and submit any progress I make so others can enjoy it too, but I need some pointers to get me started.

**Comment by liyunfan1223** [comment]
Glad to hear that! I think the bwl strategy can be a reference point. Naxx strategy is complicated because it relies on AIScripts of certain bosses, and the bot should monitor detailed events of bosses, but it is unnecessary for most of the easier dungeon fights.

About enabling the strategy automatically, it's a good idea and I will find a way to implement it later.

I understand that it's a challenge to start and am glad to provide any kind of help. You can also start by modifying strategy for different classes, like `FrostMageStrategy.cpp`, which is not fully implemented. You can observe how your modifications on the bot's behaviors more intuitively.

**Comment by noisiver** [comment]
Thank you for giving me some pointers! Most dungeons can be done without any specific strategy for sure but some really need it and that's what I'd like to try to do. I am looking through the strategies for naxx and bwl to learn the system but didn't realize I could also look at the class-specific strategies.

I'll start with something easy like the frost tombs and work my way from there. Ingvar in Utgarde Keep though, I'm not sure if that would require me to get anything from the boss script or if it could be done without it. Tank should be the only one who stands in front at any time but when he casts smash the tank should move behind him too. If I don't check to make sure they're fighting Ingvar won't they always have that strategy active regardless of what they're fighting?

**Comment by noisiver** [comment]
Just an update.

I managed to get dungeon strategies to.. kind of work. I didn't realize I actually needed a trigger for each action but now it does trigger but.. well.. Prince Keleseth, when he spawns frost tombs on people the bots do react but they basically dance in place switching between frost tomb and prince keleseth constantly so they're stuck in an infinite loop until the frost tomb goes away.

The thing is, most of this code is copy-pasted from bwl or naxx but modified to fit with this so I really don't understand why it's acting like this.

I can see when debugging that they're switching their target to frost tomb and then instantly back to boss and they loop like that over and over again until the frost tomb is gone. I'm going to step away from this for a little bit so I can clear my head and hopefully I can find a solution to the problem.

**Comment by liyunfan1223** [comment]
Of course, I haven't checked your implementation, so it's possible that the issue may be caused by other reasons that haven't been considered.

**Comment by noisiver** [comment]
I've been changing the code so many times I don't even remember what it looked like originally. This is the version I tried last and the version I have saved, although they all made the bots behave exactly the same.

Action:
```cpp
bool KelesethChooseTargetAction::Execute(Event /*event*/)
{
    GuidVector targets = context->GetValue<GuidVector>("possible targets")->Get();
    Unit* target = nullptr;

    for (GuidVector::iterator i = targets.begin(); i != targets.end(); ++i)
    {
        if (Unit* unit = botAI->GetUnit(*i))
        {
            if (unit->IsAlive() && botAI->EqualLowercaseName(unit->GetName(), "frost tomb"))
            {
                target = unit;
            }
        }
    }

    if (!target || context->GetValue<Unit*>("current target")->Get() == target)
    {
        return false;
    }

    return Attack(target);
}
```

Trigger:
```cpp
bool KelesethChooseTargetTrigger::IsActive()
{
    GuidVector targets = context->GetValue<GuidVector>("possible targets")->Get();

    for (GuidVector::iterator i = targets.begin(); i != targets.end(); ++i)
    {
        if (Unit* unit = botAI->GetUnit(*i))
        {
            if (unit->IsAlive() && botAI->EqualLowercaseName(unit->GetName(), "frost tomb") && !botAI->IsMainTank(bot))
            {
                return true;
            }
        }
    }

    return false;
}
```

Just to be clear, I've tried so many different ways of doing this and this is just the last one I tried with the least amount of code I could think of. I tried including the boss as a target if there is no frost tomb thinking maybe they are confused because of that but that didn't help either.

**Comment by noisiver** [comment]
I'm curious and maybe you can answer this, being as you're the main developer. Do the bots prioritize marked targets? Adding skull to a mob seems to make them focus on that one when possible. That's one of my questions. The second one is if this is the case then couldn't the same system be used to let strategies automate the process instead of having to put markers on everything?

**Comment by liyunfan1223** [comment]
Keep it simple, it does not apply to some special units currently, such as the frost tomb we memtioned, only to normal units. I think it's an important feature, but it may involve a series of modifications. Target selection is a bit complicated, and it can easily lead to bugs that cause the bot to get stuck. So I'll consider it carefully before implementing.

I agree that reusing priority targets can be a good implementation of these strategies.
